### PR TITLE
Robbie le pommie handle somerace conditions

### PR DESCRIPTION
**Pull request by RobbieLePommie**
As posted in https://github.com/DarkRiftNetworking/DarkRift/issues/146

### CODE REVIEW

File: DarkRift.Client/BichannelClientConnection.cs [hunk 1]
```diff
@@ -475,7 +475,9 @@ private int ProcessHeader(SocketAsyncEventArgs args)
         {
             MessageBuffer headerBuffer = (MessageBuffer)args.UserToken;
 
-            int bodyLength = BigEndianHelper.ReadInt32(headerBuffer.Buffer, headerBuffer.Offset);
+            int bodyLength = 0;
+            if (headerBuffer.Buffer?.Length > 3)
```
Comment (JamJar00) [inline DarkRift.Client/BichannelClientConnection.cs:479]: If we've got to `ProcessHeader` we aleady know we have 4 bytes so this is redundant

File: DarkRift.Server/ClientManager.cs [hunk 2]
```diff
@@ -321,6 +319,10 @@ internal void HandleNewConnection(NetworkServerConnection connection)
                     },
                     (_) => client.StartListening()
                 );
+            } else
+            {
+                logger.Info($"Handler for client [{client.ID}] is null. Enabling listener, but nothing is likely to happen.");
```
Comment (JamJar00) [inline DarkRift.Server/ClientManager.cs:324]: I like the addition of a warning here, it seems like a pretty silly thing for a user to be not doing!

Can we:
- Change this to a `warning` type log
- Change the wording to something a little more user-centric like 'Client [id] has connected but no plugin was subscribed to the ClientConnected event to process it. The client will still be connected'?
- Sort the formatting of the else keyword so it's on it's own line

Things like this are nice additions!

File: DarkRift.Server/Plugins/Listeners/Bichannel/BichannelServerConnection.cs [hunk 3]
```diff
@@ -400,7 +400,9 @@ private int ProcessHeader(SocketAsyncEventArgs args)
         {
             MessageBuffer headerBuffer = (MessageBuffer)args.UserToken;
 
-            int bodyLength = BigEndianHelper.ReadInt32(headerBuffer.Buffer, headerBuffer.Offset);
+            int bodyLength = 0;
+            if (headerBuffer.Buffer?.Length > 3)
```
Comment (JamJar00) [inline DarkRift.Server/Plugins/Listeners/Bichannel/BichannelServerConnection.cs:404]: As before

**Review by JamJar00 — CHANGES_REQUESTED** [review]

File: DarkRift.Server/ClientManager.cs (hunk 2 above)
Comment (RobbieLePommie) [reply DarkRift.Server/ClientManager.cs:324]: Ironically we had another developer encounter this yesterday, so I updated to make it clearer the plugin was not compiled. I'll amend when I next load the branch.

Comment (RobbieLePommie) [reply DarkRift.Server/ClientManager.cs:324]: Just getting back to this: by "fixing" this, it has revealed another issue. If you have a plugin that has not yet loaded, but the app is listening, then a client can still connect but the plugin will not collect the messages still loading). If that happens, it appears impossible for a client to actually connect as the connection is "open". 

Solutions are:
 - ensure the plugin's listeners are loaded immediately, and queue messages in the plugin. Not ideal as it's still a race condition.

**Comment by 4real** [comment]
I added a commit. I can't change the compare branch on this PR for some reason so I'm moving over to #163

**Pull request closed without merging**